### PR TITLE
Implemented onClose parameter type validation

### LINKED ISSUE #248: WebSocket: Validate for parameters onClose lifecycle method
#### Description:
Currently, if a user adds a non-valid parameter to methods annotated with `@onClose` to classes annotated with `@ServerEndpoint` or `@ClientEndpoint`, it will create a server error, leaving the server unresponsive. However, the user is not notified of this in the IDE. Only a small error shows up in the terminal.

For this reason, we need to apply a validation of parameters for the `onClose` lifecycle method.

#### Specification:
Valid parameters are defined as "optional Session parameter, optional EndpointConfig parameter, and zero to n parameters (of type String, any Java primitive type or boxed version thereof) annotated with a @PathParam annotation as parameters" per JakartaEE specs found [here](https://jakarta.ee/specifications/websocket/2.0/websocket-spec-2.0.html#onclose)

##### Acceptable
```java
@OnClose
public void validCloseHandler(CloseReason closeReason, Session session) {
    System.out.println("Program requested " + message + " using " + session.getId());
    session.getAsyncRemote().sendText("Lucky says hi");
}
```

##### Error
```java
@OnClose
public void notValidCloseHandler(Session session, String invalidParameter) {
   System.out.println("Program requested " + invalidParameter + " using " + session.getId());
   session.getAsyncRemote().sendText("I am pong");
}
```

#### Type of language feature proposed:
_Select all that apply_
- [x] diagnostic
- [ ] quick-fix
- [ ] snippet
- [ ] other, please specify:

### PR DESCRIPTION
**Pull request by dalidia**
Resolves #248 

Example of functionality
Screen Shot 2022-03-12 at 3 45 50 PM

### CODE REVIEW

File: jakarta.jdt/org.eclipse.lsp4jakarta.jdt.core/src/main/java/org/eclipse/lsp4jakarta/jdt/core/websocket/WebSocketDiagnosticsCollector.java [hunk 1]
```diff
@@ -310,4 +312,10 @@ private boolean isWrapper(String valueClass) {
 
         return wsEndpoint;
     }
+
+
+    private String createParamTypeDiagMsg(Set<String> methodParamOptTypes) {
+        String paramMessage = String.join("\n- ", methodParamOptTypes);
+        return "Invalid parameter type. Parameter must be of type: \n- " + paramMessage + "\n- annotated with @PathParams and of type String or any Java primitive type or boxed version thereof";
```
Comment (TheAvi123) [inline jakarta.jdt/org.eclipse.lsp4jakarta.jdt.core/src/main/java/org/eclipse/lsp4jakarta/jdt/core/websocket/WebSocketDiagnosticsCollector.java:319]: Might be a good idea to take the two parts of this message and make them constants.

**Review by TheAvi123 — COMMENTED** [review]
Looks good! Just one minor suggestion.

File: jakarta.jdt/org.eclipse.lsp4jakarta.jdt.core/src/main/java/org/eclipse/lsp4jakarta/jdt/core/websocket/WebSocketDiagnosticsCollector.java (hunk 1 above)
Comment (dalidia) [reply jakarta.jdt/org.eclipse.lsp4jakarta.jdt.core/src/main/java/org/eclipse/lsp4jakarta/jdt/core/websocket/WebSocketDiagnosticsCollector.java:319]: will do that, thanks!

**Comment by kathrynkodama** [comment]
@turkeylurkey

**Review by TheAvi123 — APPROVED** [review]

File: jakarta.jdt/org.eclipse.lsp4jakarta.jdt.core/src/main/java/org/eclipse/lsp4jakarta/jdt/core/websocket/WebSocketConstants.java [hunk 2]
```diff
@@ -60,15 +62,24 @@
...
     
-    public final static Set<String> ON_OPEN_PARAM_OPT_TYPES= new HashSet<>(Arrays.asList("jakarta.websocket.EndpointConfig", "jakarta.websocket.Session"));
-    public final static Set<String> RAW_ON_OPEN_PARAM_OPT_TYPES= new HashSet<>(Arrays.asList("EndpointConfig", "Session"));
-    
-    public final static Set<String> RAW_WRAPPER_OBJS = new HashSet<>(Arrays.asList("String", "Boolean", "Integer", "Long", "Double", "Float"));
+    public static final Set<String> RAW_WRAPPER_OBJS = new HashSet<>(Arrays.asList("String", "Boolean", "Integer", "Long", "Double", "Float"));
     public static final Set<String> WRAPPER_OBJS = RAW_WRAPPER_OBJS.stream().map(raw -> "java.lang.".concat(raw)).collect(Collectors.toSet());
+
+    // Messages
+    public static final String INITIAL_DIAG_MSG = "Invalid parameter type. Parameter must be of type: \n- ";
```
Comment (turkeylurkey) [inline jakarta.jdt/org.eclipse.lsp4jakarta.jdt.core/src/main/java/org/eclipse/lsp4jakarta/jdt/core/websocket/WebSocketConstants.java:83]: When I see this message I'm not 100% sure why. Suggestion: "Invalid parameter type. When using @OnOpen parameter must be..." or @OnClose as appropriate. Consider using String.format() so the message can be "words %s more words %s" . You can see the message as a whole and just plug in the missing words.

Comment (dalidia) [reply jakarta.jdt/org.eclipse.lsp4jakarta.jdt.core/src/main/java/org/eclipse/lsp4jakarta/jdt/core/websocket/WebSocketConstants.java]: Yeah, makes sense, I just fixed that and pushed a new commit. Thanks!

**Review by giancarlopernudisegura — APPROVED** [review]

**Review by turkeylurkey — APPROVED** [review]

**Review by kathrynkodama — APPROVED** [review]
Looks good to me @dalidia, one minor comment suggestion. Thank you!

**Review by kathrynkodama — APPROVED** [review]

**Pull request merged**